src/librustc_mir/interpret/intern.rs

```diff
@@ -328,10 +336,29 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            match base_intern_mode {
-                InternMode::Static => {}
-                InternMode::Const | InternMode::ConstBase => {
-                    // If it's not a static, it *must* be immutable.
+            match intern_kind {
+                // Mutable statics may contain mutable allocations even behind relocations
+                InternKind::Static(hir::Mutability::Mut) => {}
+                // Once we get heap allocations we need to revisit whether immutable statics can
+                // refer to mutable (e.g. via interior mutability) allocations.
+                // Note: this is never the base value of the static, we can only get here for
+                // pointers encountered inside the base allocation, and then only for ones not at
+                // reference type, as that is checked by the type based main interner.
+                InternKind::Static(hir::Mutability::Not) => {
+                    alloc.mutability = Mutability::Not;
+                }
+                // Raw pointers in promoteds may only point to immutable things so we mark
+                // everything as immutable. Creating a promoted with interior mutability is UB, but
+                // there's no way we can check whether the user is using raw pointers correctly.
+                // So all we can do is mark this as immutable here.
+                // It is UB to mutate through a raw pointer obtained via an immutable reference.
+                // Since all references and pointers inside a promoted must by their very definition
+                // be created from an immutable reference, mutating though them would be UB.
+                InternKind::Promoted => {
+                    alloc.mutability = Mutability::Not;
+                }
+                InternKind::Constant | InternKind::ConstProp => {
```

Comment:
Right, I was thinking about this branch here when asking about nested allocations in constants. Isn't this unreachable then?

Reply:
It is reachable, but I'm entirely unclean on why it is reachable

```rust
const RAW_TRAIT_OBJ_CONTENT_INVALID: *const dyn Trait = &unsafe { BoolTransmute { val: 3 }.bl } as *const _;
```
hits the `InternKind::Constant` arm.

See the MIR on https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3bc6c8b165b614bbe6cf5566d30752de there's no `StorageDead` for the allocation containing the `bool` with bit pattern `3`.

I'm not really sure what's going on there. Technically, since we're inside a constant, shouldn't this trigger promotion and promote the transmuted value?

Reply:
Under unleash it's also reached in

```rust
const MUTABLE_BEHIND_RAW: *mut i32 = &UnsafeCell::new(42) as *const _ as *mut _;
```

but that triggers the delay_span_bug as expected.

Also under unleash, but without an ICE (just triggering the dynamic checks in interp):

```rust
const MUTABLE_BEHIND_RAW: *mut i32 = &UnsafeCell::new(42) as *const _ as *mut _;
//~^ WARN: skipping const checks

const MUTATING_BEHIND_RAW: () = {
    // Test that `MUTABLE_BEHIND_RAW` is actually immutable, by doing this at const time.
    unsafe {
        *MUTABLE_BEHIND_RAW = 99 //~ ERROR any use of this value will cause an error
    }
};
```

Reply:
> I'm not really sure what's going on there. Technically, since we're inside a constant, shouldn't this trigger promotion and promote the transmuted value?

Ah, this must be the "not-promotion" also described [in this document](https://github.com/rust-lang/const-eval/blob/master/promotion.md) (Ctrl-F "looks like"). That makes sense.

Reply:
What I don't understand though is why `RAW_TRAIT_OBJ_CONTENT_INVALID` doesn't trigger the ICE. Why is the allocation already immutable?

Reply:
The `&unsafe { BoolTransmute { val: 3 }.bl }` part is actually promoted

```rust
const  RAW_TRAIT_OBJ_CONTENT_INVALID: *const dyn Trait = {
    let mut _0: *const dyn Trait;        // return place in scope 0 at src/lib.rs:1:38: 1:54
    let mut _1: *const dyn Trait;        // in scope 0 at src/lib.rs:1:57: 1:108
    let mut _2: *const bool;             // in scope 0 at src/lib.rs:1:57: 1:96
    let _3: &bool;                       // in scope 0 at src/lib.rs:1:57: 1:96
    scope 1 {
    }

    bb0: {
        StorageLive(_1);                 // bb0[0]: scope 0 at src/lib.rs:1:57: 1:108
        StorageLive(_2);                 // bb0[1]: scope 0 at src/lib.rs:1:57: 1:96
        StorageLive(_3);                 // bb0[2]: scope 0 at src/lib.rs:1:57: 1:96
        _3 = &(promoted[0]: bool);       // bb0[3]: scope 0 at src/lib.rs:1:57: 1:96
        _2 = &raw const (*_3);           // bb0[4]: scope 0 at src/lib.rs:1:57: 1:96
        _1 = move _2 as *const dyn Trait (Pointer(Unsize)); // bb0[5]: scope 0 at src/lib.rs:1:57: 1:96
        _0 = _1;                         // bb0[6]: scope 0 at src/lib.rs:1:57: 1:108
        StorageDead(_3);                 // bb0[7]: scope 0 at src/lib.rs:1:107: 1:108
        StorageDead(_1);                 // bb0[8]: scope 0 at src/lib.rs:1:107: 1:108
        return;                          // bb0[9]: scope 0 at src/lib.rs:1:1: 1:109
    }
}

promoted[0] in  RAW_TRAIT_OBJ_CONTENT_INVALID: bool = {
    let mut _0: bool;                    // return place in scope 0 at src/lib.rs:1:57: 1:96
    let mut _1: bool;                    // in scope 0 at src/lib.rs:1:58: 1:96
    let mut _2: BoolTransmute;           // in scope 0 at src/lib.rs:1:67: 1:91
    scope 1 {
    }

    bb0: {
        (_2.0: u8) = const 3u8;          // bb0[0]: scope 0 at src/lib.rs:1:67: 1:91
                                         // ty::Const
                                         // + ty: u8
                                         // + val: Value(Scalar(0x03))
                                         // mir::Constant
                                         // + span: src/lib.rs:1:88: 1:89
                                         // + literal: Const { ty: u8, val: Value(Scalar(0x03)) }
        _1 = (_2.1: bool);               // bb0[1]: scope 0 at src/lib.rs:1:67: 1:94
        _0 = move _1;                    // bb0[2]: scope 0 at src/lib.rs:1:57: 1:96
        return;                          // bb0[3]: scope 0 at src/lib.rs:1:57: 1:96
    }
}
```

Reply:
> The &unsafe { BoolTransmute { val: 3 }.bl } part is actually promoted

But then why is this arm (`InternKind::Constant | InternKind::ConstProp`) reached? Your answers just keep opening new questions.^^ If you have a holistic view of what happens, it would really help to write that down. I am trying to prove that with my questions, but each answer just exposes the next layer in what seems to be a maze...

Reply:
(The PR doesn't change anything here so this doesn't block landing, but I'd really like to understand this and then see it documented.)

Reply:
I would find documentation here immensely helpful.

Reply:
> If you have a holistic view of what happens, it would really help to write that down.

I don't, which is why my answers are so confusing. They are just a brain dump of me discovering what is going on, not me understanding it entirely

Reply:
These issues are preexisting, so after this PR is merged I'll open a new one to write docs and actually figure out what is going on

Reply:
> I don't, which is why my answers are so confusing. They are just a brain dump of me discovering what is going on, not me understanding it entirely

Fair enough. :)